Fix dataset caching in contexts based on transaction timeout and runtime...

ParthGandhi opened the pull request:
... args

Fixes for: https://issues.cask.co/browse/CDAP-1211 and https://issues.cask.co/browse/CDAP-1212

Add the hashcode of the arguments to the key used to cache datasets and extend the cache timeout to be greater than the transaction timeout.

albertshau commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetContext.java:
```diff
@@ -135,15 +137,16 @@ public DynamicDatasetContext(TransactionContext context, DatasetFramework datase
       dsArguments.putAll(arguments);
     }
 
+    String datasetKey = String.format("%s.%s", name, dsArguments.hashCode());
```
I think it would be cleaner to have a dataset key class with the 2 fields, which can be internal to this class, and use that as the map key instead of formatting the string.

poornachandra replied:
+1

ParthGandhi replied:
Will do. Though needs to be a public class since both `BasicNotificationContext` and `BasicServiceWorkerContext` need access to it since they define the cache.

albertshau commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetContext.java:
```diff
@@ -28,22 +28,24 @@
   private final DatasetFramework datasetFramework;
   private final ClassLoader classLoader;
   private final Map<String, String> runtimeArguments;
+  private final Set<String> txnInProgressDatasets = Sets.newHashSet();
```
should initialize this in the constructor with whatever is already in the transaction context.

poornachandra replied:
I don't think it is required since a new transaction will be created for this object.

albertshau replied:
but it is possible to add txawares to the context before passing it in here.

poornachandra replied:
Ah - I missed the part that txcontext is passed to this class.

albertshau replied:
as discussed, instances created by this context are separate, so ignore my comment

jguery commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/BasicServiceWorkerContext.java:
```diff
@@ -95,10 +96,17 @@ public BasicServiceWorkerContext(ServiceWorkerSpecification spec, Program progra
+
+    // The cache expiry should be greater than (2 * transaction.timeout) and at least 2 minutes.
+    // This ensures that when a dataset instance is requested multiple times during a single transaction,
+    // the same instance is always returned.
```
Not unless a dataset is cached in a first transaction, and then in the course of another transaction, it could get expired. Right? Or is this ``BasicServiceWorkerContext`` only meant to live for the course of one transaction only?

ParthGandhi replied:
The context lives across multiple transactions. However, I'm not sure what you mean. If it expires during the second transaction a new instance is given - which is not a problem. A problem arises only when 2 different instance of the same dataset are given during the same transaction.

Unless I misunderstand what you're saying?

poornachandra commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetContext.java:
```diff
@@ -154,8 +157,11 @@ public DynamicDatasetContext(TransactionContext context, DatasetFramework datase
-      if (dataset instanceof TransactionAware) {
+      // For every instance of a TransactionAware dataset acquired, add it to the
+      // current transaction only if it hasn't been added previously.
+      if (!txnInProgressDatasets.contains(datasetKey) && dataset instanceof TransactionAware) {
```
It would be more readable to flip the condition.

ParthGandhi replied:
Did it this way since if the set contains that key it implicitly implies that a dataset is TxAware, so you avoid the additional check. But I agree readability trumps that.

ParthGandhi commented:
@albertshau @poornachandra Addressed comments. Please review.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DatasetCacheKey.java:
```diff
@@ -0,0 +1,55 @@
+  private final String name;
+  private final Map<String, String> arguments;
+
+  DatasetCacheKey(String name, Map<String, String> arguments) {
```
Why package only constructor for a public class? Seems to me this class could be just an protected static inner class inside `DynamicDatasetContext`?

ParthGandhi replied:
@chtyim The class is used in `BasicServiceWorkerContext` so it needs to be public and can't be protected inside `DynamicDatasetContext`. Will make the constructor public though.

poornachandra commented on cdap-unit-test/src/test/java/co/cask/cdap/test/app/AppWithServices.java:
```diff
@@ -243,6 +246,13 @@ public void stop() {
+
+            // Get another instance of the dataset with different runtime arguments and write a new value
+            // only if its a different instance
+            KeyValueTable table2 = context.getDataset(DATASET_NAME, ImmutableMap.of("arg", "value"));
```
It would be a good idea to get the Dataset again with the same arguments to make sure you get the same Dataset.

poornachandra commented:
Minor comment, otherwise LGTM :+1:

ParthGandhi commented:
@chtyim Addressed your comment. 
@poornachandra Can you make sure https://github.com/caskdata/cdap/commit/df50a2d3e6908cc4308f75f83d5354480c3348e4 covers every case we discussed? I believe it does.

poornachandra commented:
LGTM :+1:

ParthGandhi commented:
https://builds.cask.co/browse/CDAP-DUT566-8
